dkg-gadget/src/worker.rs

```diff
@@ -478,42 +493,168 @@ where
+				let is_ready_to_vote = next_rounds.is_ready_to_vote();
 
 				if is_ready_to_vote && self.queued_keygen_in_progress {
-					debug!(target: "dkg", "🕸️  Queued DKGs keygen has completed");
+					info!(target: "dkg", "🕸️  queued_keygen: Queued DKGs keygen has completed");
 					self.queued_keygen_in_progress = false;
-					let pub_key = self
-						.next_rounds
-						.as_mut()
-						.unwrap()
+					let pub_key = next_rounds
 						.get_public_key()
 						.unwrap()
 						.get_element()
 						.serialize_uncompressed()
 						.to_vec();
-					let offchain = self.backend.offchain_storage();
 
-					if let Some(mut offchain) = offchain {
-						offchain.set(STORAGE_PREFIX, OFFCHAIN_PUBLIC_KEY, &pub_key);
-					}
+					self.gossip_public_key(pub_key, dkg_msg.round_id);
 				}
+
+				self.next_rounds = Some(next_rounds);
 			}
 		}
 
+		self.handle_public_key_broadcast(dkg_msg.clone());
+
 		self.send_outgoing_dkg_messages();
 
 		self.process_finished_rounds();
 	}
 
+	/// Offchain features
+
+	fn listen_and_clear_offchain_storage(&mut self, header: &B::Header) {
+		let at = BlockId::hash(header.hash());
+		let next_dkg_public_key = self.client.runtime_api().next_dkg_pub_key(&at).ok();
+
+		if next_dkg_public_key.is_some() {
+			let offchain = self.backend.offchain_storage();
+			if let Some(mut offchain) = offchain {
+				if offchain.get(STORAGE_PREFIX, AGGREGATED_PUBLIC_KEYS).is_some() {
+					offchain.remove(STORAGE_PREFIX, AGGREGATED_PUBLIC_KEYS);
+
+					offchain.remove(STORAGE_PREFIX, SUBMIT_KEYS_AT);
+
+					self.dkg_state.listening_for_pub_key = false;
+					self.aggregated_public_keys.keys_and_signatures = vec![];
+				}
+			}
+		}
+	}
+
+	fn gossip_public_key(&mut self, public_key: Vec<u8>, round_id: RoundId) {
+		let public = self
+			.key_store
+			.authority_id(&self.key_store.public_keys().unwrap())
+			.unwrap_or_else(|| panic!("Halp"));
+
+		if let Ok(signature) = self.key_store.sign(&public, &public_key) {
+			let encoded_signature = signature.encode();
+			let message = DKGMessage::<AuthorityId, DKGPayloadKey> {
+				id: public.clone(),
+				round_id,
+				payload: DKGMsgPayload::PublicKeyBroadcast(DKGPublicKeyMessage {
+					pub_key: public_key.clone(),
+					signature: encoded_signature.clone(),
+				}),
+			};
+
+			self.gossip_engine
+				.lock()
+				.gossip_message(dkg_topic::<B>(), message.encode(), true);
+
+			self.aggregated_public_keys
+				.keys_and_signatures
+				.push((public_key.clone(), encoded_signature));
+			info!(target: "dkg", "queued_keygen: gossiping local node  {:?} public key and signature", public)
+		}
+		else {
+			error!(target: "dkg", "queued_keygen: Could not sign public key");
+		}
+	}
+
+	fn handle_public_key_broadcast(&mut self, dkg_msg: DKGMessage<Public, DKGPayloadKey>) {
+		if !self.dkg_state.listening_for_pub_key {
+			return
+		}
+
+		match dkg_msg.payload {
+			DKGMsgPayload::PublicKeyBroadcast(msg) => {
+				let key_and_sig = (msg.pub_key, msg.signature);
```

Comment:
We have to validate that this key is in fact a DKG authority / validator / collator. We run the risk of being open to spoofing attacks from non-authorities on the gossip network.